File: python/2_Add_Two_Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode()
        ptr = head
        carry = 0
        
        while l1 or l2:
            if l1:
                n1 = l1.val
                l1 = l1.next
            else:
                n1 = 0
            
            if l2:
                n2 = l2.val
                l2 = l2.next
            else:
                n2 = 0
        
            tmpNode = ListNode((n1+n2+carry) % 10)
            carry = (n1+n2+carry) // 10
            ptr.next = tmpNode
            ptr = ptr.next
        
        if carry != 0:
            tmpNode = ListNode(carry)
            ptr.next = tmpNode
        
        return head.next
            
            
        

File: c++/2_Add-Two-Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void add_recursion(ListNode* l1, ListNode* l2, ListNode* curNode, int increment) {
        if(l1==nullptr && l2==nullptr && increment==0) return;
        int num = increment;
        if(l1!=nullptr){
            num+=l1->val;
            l1=l1->next;
        }
        if(l2!=nullptr){
            num+=l2->val;
            l2=l2->next;
        }
        
        ListNode* temp = new ListNode(num%10);
        curNode->next = temp;
        add_recursion(l1, l2, curNode->next, num/10);
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode();
        ListNode* curNode = head;
        add_recursion(l1, l2, curNode, 0);
        return head->next;
    }
};